extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/NotifyingUserTransaction.java

```diff
@@ -0,0 +1,68 @@
+package io.quarkus.narayana.jta.runtime;
+
+import javax.transaction.HeuristicMixedException;
+import javax.transaction.HeuristicRollbackException;
+import javax.transaction.NotSupportedException;
+import javax.transaction.RollbackException;
+import javax.transaction.SystemException;
+import javax.transaction.UserTransaction;
+
+import org.jboss.logging.Logger;
+
+public class NotifyingUserTransaction extends TransactionScopedNotifier implements UserTransaction {
+
+    private static final Logger LOG = Logger.getLogger(NotifyingUserTransaction.class);
+
+    private final UserTransaction delegate;
+
+    public NotifyingUserTransaction(UserTransaction delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public void begin() throws NotSupportedException, SystemException {
+        delegate.begin();
+        initialized(delegate);
```

Comment:
I'm not so sure about the payload used. The original `CDIDelegatingTransactionManager` used the `javax.transaction.Transaction` and for the `@Destroyed` event the value of `this.toString()`. We should probably follow the same strategy...

Reply:
I've replaced the `this.toString()` with `this`.

Reply:
I'd go with what's directly in Narayana, as seen [here](https://github.com/jbosstm/narayana/blob/master/ArjunaJTA/cdi/classes/com/arjuna/ats/jta/cdi/NarayanaTransactionManager.java#L76-L97).
JTA spec doesn't seem concerned about this at all so it's just about not polluting other observer methods for which it is reasonable to use either JTA-specific class (`Transaction`) or plain `Object`.

Reply:
I tried looking at the spec for a definitive answer for what the initialised and destroyed event payload should contain but the spec leaves it unspecified.

TBH if I was implementing it I'd put the narayana transaction id in the payload (for the initialized, beforeDestroyed and destroyed events).

The Quarkus implementation is using delegate which is an instance of `UserTransaction` which in the Narayana case delegates to our implementation, therefore the `toString()` method has no relationship to the actual transaction scope (the actual "scope" is the transaction that is bound to the current thread - except that there would be no actual transaction for the destroyed event). It is because of these ambiguities that I would just use the transaction id which the listener can use for all transaction scope events in order to uniquely identify the scope/transaction.

Reply:
You can construct the payload in a similar way that you do in `NotifyingTransactionManager`: 

By calling `getTransaction()` on the transaction manager you will get the current transaction bound to the calling thread and that object will uniquely correspond to the active transaction scope. In the commit method you will need to cache the transaction before calling `delegate.commit();` since the call may return null after the transaction gets committed. Similarly for rollback. You should be able to use that object directly in the payload since its' `toString()` method will be unique to the transaction or you can just use the result of calling `toString()` as the payload.

Reply:
Another observation is that in `NotifyingTransactionManager#commit()` we use `this.getTransaction()` for the payload in the beforeDestroyed event but we use `destroyed(this);` for the payload in the destroyed event. If the payload is meant to be the same (which I think it should) then I think the call to `this.getTransaction()` should be cached and the same object used for both events

Reply:
> If the payload is meant to be the same (which I think it should) then I think the call to this.getTransaction() should be cached and the same object used for both events

It doesn't have to be the same. And again, I was inspired by the original code: https://github.com/quarkusio/quarkus/blob/2.13.3.Final/extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/CDIDelegatingTransactionManager.java#L97-L111. My guess that the `this.getTransaction()` wasn't used because it's not usable after `commit()/rollback()`?

In any case, I agree that a transaction id makes more sense for all events.

Reply:
> My guess that the `this.getTransaction()` wasn't used because it's not usable after `commit()/rollback()`?

FTR I've found this in the javadoc of `UserTransaction#commit()`: _"When this method completes, the thread is no longer associated with a transaction."_